tools/meteor-services/deploy.js

```diff
@@ -799,6 +814,86 @@ var listSites = function () {
   return 0;
 };
 
+// Given a hostname, add "http://" or "https://" as
+// appropriate. (localhost gets http; anything else is always https.)
+function addScheme(hostOrURL) {
+  if (hostOrURL.match(/^http/)) {
+    return hostOrURL;
+  } else if (hostOrURL.match(/^localhost(:\d+)?$/)) {
+    return "http://" + hostOrURL;
+  } else {
+    return "https://" + hostOrURL;
+  }
+};
+
+// Maps from "site" to Promise<deploy URL>, so we don't have to re-ping on each
+// RPC (even if the calls to getDeployURL overlap).
+const galaxyDiscoveryCache = new Map;
+
+// getDeployURL returns the a Promise for the base deploy URL for the given app.
+// "app" may be falsey for certain RPCs (eg meteor list-sites).
+function getDeployURL(site) {
+  // Always trust explicitly configuration via env.
+  if (process.env.DEPLOY_HOSTNAME) {
+    return Promise.resolve(addScheme(process.env.DEPLOY_HOSTNAME));
+  }
+
+  const defaultURL = "https://galaxy.meteor.com";
+
+  // No site? Just use the default.
+  if (!site) {
+    return Promise.resolve(defaultURL);
+  }
+
+  // If we have a site, we can try to do Galaxy discovery.
+
+  // Do we already have an answer?
+  if (galaxyDiscoveryCache.has(site)) {
+    return galaxyDiscoveryCache.get(site);
+  }
+
+  // Otherwise, try https first, then http, then just use the default.
+  const p = discoverGalaxy(site, "https")
+          .catch(() => discoverGalaxy(site, "http"))
+          .catch(() => defaultURL);
+  galaxyDiscoveryCache.set(site, p);
+  return p;
+}
+
+// discoverGalaxy returns a promise which resolves to the URL to use for Galaxy
+// discovery, or an error if it couldn't be fetched.
+function discoverGalaxy(site, scheme) {
```

Comment:
Even better: make `discoverGalaxy` an `async` function, and then just `throw` or `return` instead of creating a `new Promise` and calling `reject` or `resolve`.